### PR TITLE
NODE-2499: upgrade to bson v4

### PR DESCRIPTION
**Pull request by mbroadst**
This upgrades the C++ addon to use BSONv4 for the upcoming v4
driver release.

NODE-2499

### CODE REVIEW

File: test/utils.js [hunk 1]
```diff
@@ -1,7 +1,5 @@
 var BSON = require('..');
 
 module.exports = function createBSON() {
-  return new BSON([BSON.Binary, BSON.Code, BSON.DBRef, BSON.Decimal128,
-    BSON.Double, BSON.Int32, BSON.Long, BSON.Map, BSON.MaxKey, BSON.MinKey,
-    BSON.ObjectId, BSON.BSONRegExp, BSON.Symbol, BSON.Timestamp]);
+  return BSON;
```
Comment (nbbeeken) [inline test/utils.js]: Is it worth finding and dropping usages of this helper?

File: test/node/bson_test.js [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
-  Double = BSON.Double,
-  MinKey = BSON.MinKey,
-  MaxKey = BSON.MaxKey,
-  BinaryParser = require('../binary_parser').BinaryParser,
-  vm = require('vm');
-
-var createBSON = require('../utils');
+const fs = require('fs');
+const { Buffer } = require('buffer');
+const { expect } = require('chai');
```
Comment (nbbeeken) [inline test/node/bson_test.js]: Is this bump going to drop node 4 support? (see: travis errors)

Comment (mbroadst) [reply test/node/bson_test.js]: Yes it will. `bson-ext` v4 will track `bson` v4, and have the same version requirements.

**Comment by bompus** [comment]
Any ETA on this? I'm curious to see the benchmark against js-bson 4.1.0

**Comment by mbroadst** [comment]
@bompus the main bson module is undergoing a decent amount of surgery right now to convert it to TypeScript, so there is no target date for the `bson-ext` 4.0 release. You should be able to take this branch and benchmark, however.

**Comment by nbbeeken** [comment]
Resolved by #67.

The benchmark script was updated to work with the changes made there. We plan to release a new major version of bson-ext soon. You can checkout the 4.0 branch of this repo and run:

```sh
node benchmarks/serialization_benchmark.js
```

**Pull request closed without merging**